[TEP-0089] SPIRE for non-falsifiable provenance. (#6160)
tektoncd/pipeline: docs/spire.md

-->
⚠️ This is a work in progress: SPIRE support is not yet functional

> TaskRun result attestations is currently an alpha experimental feature. Currently all that is implemented is support for configuring Tekton to connect to SPIRE and enabling TaskRun to sign and verify the TaskRunStatus. See TEP-0089 for details on the overall design and feature set.

Reviewer: Add a link for users who may not know where to find TEPs:

```suggestion
TaskRun result attestations is currently an alpha experimental feature. Currently all that is implemented is support for configuring Tekton to connect to SPIRE and enabling TaskRun to sign and verify the TaskRunStatus. See [TEP-0089] for details on the overall design and feature set.

 [TEP-0089]: https://github.com/tektoncd/community/blob/main/teps/0089-nonfalsifiable-provenance-support.md
```

Author: +1. 
Added a link to the TEP-0089 string below.

> TaskRun result attestations is currently an alpha experimental feature. Currently all that is implemented is support for configuring Tekton to connect to SPIRE and enabling TaskRun to sign and verify the TaskRunStatus. See [TEP-0089](https://github.com/tektoncd/community/blob/main/teps/0089-nonfalsifiable-provenance-support.md) for details on the overall design and feature set.

Reviewer: ```suggestion
TaskRun result attestations is currently an alpha experimental feature. Currently we support configuring Tekton to connect to SPIRE and enabling TaskRun to sign and verify the TaskRunStatus. The next and final step is to enable Chains to verify TaskRun Results. See [TEP-0089](https://github.com/tektoncd/community/blob/main/teps/0089-nonfalsifiable-provenance-support.md) for details on the overall design and feature set.
```

> This being a large feature, this will be implemented in the following phases. This document will be updated as we implement new phases.

Reviewer: should this comment be updated now that we've finished the implementation

Author: Agreed, also updated the features completed with "done" status.

1.  Add a client for SPIRE (done).
2.  Add a configMap which initializes SPIRE (in progress).
3.  Modify TaskRun to sign and verify TaskRunStatus using SPIRE.
4.  Modify Tekton Chains to verify the TaskRunStatus.

> When the TaskRun result attestations feature is enabled, all TaskRuns will produce a signature alongside its results, which can then be used to validate its provenance. For example, a TaskRun result that creates user-specified results `commit` and `url` would look like the following. `SVID`, `RESULT_MANIFEST`, `RESULT_MANIFEST.sig`, `commit.sig` and `url.sig` are generated attestations by the integration of SPIRE and Tekton Controller.

Reviewer: > TaskRun result attestations feature is enabled

which feature is this referring to? how can it be enabled? please be specific, and add a link to the feature

Author: Added a link on "enabled" to point to how to enable this feature.

$ tkn tr describe non-falsifiable-provenance
Name:              non-falsifiable-provenance
Namespace:         default
Service Account:   default
Timeout:           1m0s
Labels:
 app.kubernetes.io/managed-by=tekton-pipelines

🌡️  Status

STARTED          DURATION     STATUS
38 seconds ago   36 seconds   Succeeded

📝 Results

 NAME        VALUE
 ∙ bar       world
 ∙ foo       hello

🦶 Steps

 NAME                STATUS
 ∙ non-falsifiable   Completed
```

## How is the result being verified

The signatures are being verified by the Tekton controller, the process of verification is as follows:

- Verifying the SVID
  - Obtain the trust bundle from the SPIRE server
  - Verify the SVID with the trust bundle
  - Verify that the SVID spiffe ID is for the correct TaskRun
- Verifying the result manifest
  - Verify the content of `RESULT_MANIFEST` with the field `RESULT_MANIFEST.sig` with the SVID public key
  - Verify that there is a corresponding field for all items listed in `RESULT_MANIFEST` (besides SVID and `*.sig` fields)
- Verify individual result fields
  - For each of the items in the results, verify its content against its associated `.sig` field

# TaskRun Status attestations

> Each TaskRun status that is written by the tekton-pipelines-controller will be signed to ensure that there is no external

Reviewer: ```suggestion
Each TaskRun status that is written by the Tekton Pipelines controller will be signed to ensure that there is no external
```

> tampering of the TaskRun status. Upon each retrieval of the TaskRun, the tekton-pipelines-controller checks if the status is initialized,

Reviewer: ```suggestion
tampering of the TaskRun status. Upon each retrieval of the TaskRun, the Tekton Pipelines controller checks if the status is initialized,
```

and that the signature validates the current status.
The signature and SVID will be stored as annotations on the TaskRun Status field, and can be verified by a client.

The verification is done on every consumption of the TaskRun except when the TaskRun is uninitialized. When uninitialized, the 

> tekton-pipelines-controller is not influenced by fields in the status and thus will not sign incorrect reflections of the TaskRun.

Reviewer: ```suggestion
Tekton Pipelines controller is not influenced by fields in the status and thus will not sign incorrect reflections of the TaskRun.
```

> The spec and TaskRun annotations/labels are not signed when there are valid interactions from other controllers or users (i.e. cancelling taskrun).

Reviewer: ```suggestion
The TaskRun spec and annotations/labels are not signed when there are valid interactions from other controllers or users (i.e. cancelling taskrun).
```

      8xVWrQr8+i6yMLDm9IUjtvTbz9ofjSsWL6c/+rxmmRYCIBTiJ/HW7di3inSfxwqK
      5DKyPrKoR8sq8Ne7flkhgbkg
      -----END CERTIFICATE-----
    tekton.dev/status-hash: 76692c9dcd362f8a6e4bda8ccb4c0937ad16b0d23149ae256049433192892511
    tekton.dev/status-hash-sig: MEQCIFv2bW0k4g0Azx+qaeZjUulPD8Ma3uCUn0tXQuuR1FaEAiBHQwN4XobOXmC2nddYm04AZ74YubUyNl49/vnbnR/HcQ==
  completionTime: "2022-03-04T19:11:22Z"
  conditions:
  - lastTransitionTime: "2022-03-04T19:11:22Z"
    message: All Steps have completed executing
    reason: Succeeded
    status: "True"
    type: Succeeded
  - lastTransitionTime: "2022-03-04T19:11:22Z"
    message: Spire verified
    reason: TaskRunResultsVerified
    status: "True"
    type: SignedResultsVerified
  podName: non-falsifiable-provenance-pod
  startTime: "2022-03-04T19:10:46Z"
  steps:
  ...
  <TRUNCATED>
```

## How is the status being verified

The signature are being verified by the Tekton controller, the process of verification is as follows:

- Verify status-hash fields
  - verify `tekton.dev/status-hash` content against its associated `tekton.dev/status-hash-sig` field. If status hash does 

>     not match invalidate the `tekton.dev/verified = no` annotation will be added

Reviewer: is the `TaskRunResultsVerified` also modified?
